nengo/simulator.py

```diff
@@ -194,6 +195,24 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_value, traceback):
         self.close()
 
+    def __getstate__(self):
+        signals = {k: v for k, v in self.signals.items() if not k.readonly}
+        probe_outputs = {probe: self._probe_outputs[probe]
+                         for probe in self.model.probes}
+        return dict(model=self.model, signals=signals,
+                    probe_outputs=probe_outputs, dt=self.dt, seed=self.seed,
+                    progress_bar=self.progress_bar, optimize=self.optimize)
+
+    def __setstate__(self, state):
+        self.__init__(network=None, model=state['model'],
+                      dt=state['dt'], seed=state['seed'],
+                      progress_bar=state['progress_bar'],
+                      optimize=state['optimize'])
+        for key, value in state['signals'].items():
+            self.signals[key] = value
+        for key, value in state['probe_outputs'].items():
+            self._probe_outputs[key].extend(value)
```

Comment:
Are there any situations where the value of `self.closed` can matter?

Reply:
After looking at the main unit test, it's a little weird to me that I can't do:
```python
with sim:
    sim.run(0.001)
with sim:
    sim.run(0.001)
```
but I can if I pickle and then unpickle it in between. What was the motivation for disallowing reopening the simulator?

Reply:
On other backends, we may destroy the data when the simulator is closed.

Reply:
Why do we disallow it for the reference backend then? This allows the reference simulator to be reopened by unpickling it. So It seems that for consistency either we should disallow pickling if the reference simulator is closed, or allow the reference simulator to be reopened.

Reply:
Pickling can be used to save probe data, too. I think we should just store the value of `closed` as part of pickling, so if you pickle->unpickle a closed simulator, it's still closed.

Reply:
I think the main argument for not allowing it on the reference backend is so that users don't get into the habit of being able to do that, since if they switch simulators they can't.